### ISSUE TITLE
IPv6 port forwards do not work for link local IP addresses (from external host on same network)

### ISSUE THREAD

**Issue by chrisbraucker** [containers/netavark#745]
# Description

I am running a rootful container with `sudo` from a non-root account in an IPv6-enabled bridge network, and forward ports with the `--publish` parameter, e.g. `--publish [::]:8080:8080/tcp`.

Firewall was completely disabled while testing, and I obfuscated IP and MAC addresses. Adresses without an `aaaa` were left untouched.

I want to use link local addresses since my ISP may rotate GUA prefixes.

# Expected result

The container is accessible via its own IPv6 and due to forwarding, also on all host IPv6 addresses, both locally and from external hosts (same network).

E.g. if I run a service directly on the host, it is available to other devices on the same network via link local address, but not when run in container.

# Actual result

The container is accessible via its own IPv6 and via the host's GUA, but not via the host's link local address from external hosts on the same network.

# Steps to reproduce

Premise, host has IPv6 addresses on `eth0` with

`fe80::aaaa:fc80` (LLA) and `2a00:1e:8085:6801::aaaa:fc80` (GUA)

Create a network and container with

```bash
# creates subnet fd8a:e26d:2709:9821::/64, with eth device podman1
podman network create --ipv6 podman6
# assigns fd8a:e26d:2709:9821::3 (from podman inspect)
podman run --rm -it \
    --network podman6 \
    --publish '0.0.0.0:8080:8080/tcp' --publish '[::]:8080:8080/tcp' \
    alpine nc -v -l -p 8080
```

then send something over the wire from the host:

```bash
# arrives as expected
echo "test" | nc -v fd8a:e26d:2709:9821::3 8080
# arrives as expected
echo "test" | nc -v 2a00:1e:8085:6801::aaaa:fc80 8080
# does not arrive (as expected)
echo "test" | nc -v fe80::aaaa:fc80%eth0 8080
# arrives as expected
echo "test" | nc -v fe80::aaaa:fc80%podman1 8080
```

however, from other hosts on the same network

```bash
# does not arrive (have not set up NDP)
echo "test" | nc -v fd8a:e26d:2709:9821::3 8080
# arrives as expected
echo "test" | nc -v 2a00:1e:8085:6801::aaaa:fc80 8080
# does not arrive
echo "test" | nc -v fe80::aaaa:fc80%eth0 8080
# however ping6 does work
ping6 -c1 fe80::aaaa:fc80%eth0
PING6(56=40+8+8 bytes) fe80::aaaa:dfd5%eth0 --> fe80::aaaa:fc80%eth0
16 bytes from fe80::aaaa:fc80%eth0, icmp_seq=0 hlim=64 time=1.411 ms
  [...]
```

`tcpdump -l -e -n tcp port 8080` on the host shows SYN packets arriving from other hosts, but it seems they don't get routed to the container. Running the same command in the container namespace (`ip netns exec`) shows no packets whatsoever when external hosts try to connect.

# Introspection on host

```
$> ip6tables -L -v -n
Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    9   671 NETAVARK_FORWARD  all      *      *       ::/0                 ::/0                 /* netavark firewall plugin rules */

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain NETAVARK_FORWARD (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all      *      *       ::/0                 fd8a:e26d:2709:9821::/64  ctstate RELATED,ESTABLISHED
    0     0 ACCEPT     all      *      *       fd8a:e26d:2709:9821::/64  ::/0
```

```
$> ip6tables -L -v -n -t nat
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
   32  3325 NETAVARK-HOSTPORT-DNAT  all      *      *       ::/0                 ::/0                 ADDRTYPE match dst-type LOCAL

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
   10   800 NETAVARK-HOSTPORT-DNAT  all      *      *       ::/0                 ::/0                 ADDRTYPE match dst-type LOCAL

Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
  328 27076 NETAVARK-HOSTPORT-MASQ  all      *      *       ::/0                 ::/0
    0     0 NETAVARK-44A471C841C13  all      *      *       fd8a:e26d:2709:9821::/64  ::/0

Chain NETAVARK-44A471C841C13 (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all      *      *       ::/0                 fd8a:e26d:2709:9821::/64
    0     0 MASQUERADE  all      *      *       ::/0                !ff00::/8

Chain NETAVARK-DN-44A471C841C13 (1 references)
 pkts bytes target     prot opt in     out     source               destination
    0     0 NETAVARK-HOSTPORT-SETMARK  tcp      *      *       fd8a:e26d:2709:9821::/64  ::/0                 tcp dpt:8080
    0     0 NETAVARK-HOSTPORT-SETMARK  tcp      *      *       ::1                  ::/0                 tcp dpt:8080
   11   908 DNAT       tcp      *      *       ::/0                 ::/0                 tcp dpt:8080 to:[fd8a:e26d:2709:9821::3]:8080

Chain NETAVARK-HOSTPORT-DNAT (2 references)
 pkts bytes target     prot opt in     out     source               destination
   11   908 NETAVARK-DN-44A471C841C13  tcp      *      *       ::/0                 ::/0                 tcp dpt:8080 /* dnat name: podman6 id: feb22c9141d38686d4a7333f49b100905fa12bc09402617a4e92c8d547f67dc4 */

Chain NETAVARK-HOSTPORT-MASQ (1 references)
 pkts bytes target     prot opt in     out     source               destination
    4   320 MASQUERADE  all      *      *       ::/0                 ::/0                 /* netavark portfw masq mark */ mark match 0x2000/0x2000

Chain NETAVARK-HOSTPORT-SETMARK (2 references)
 pkts bytes target     prot opt in     out     source               destination
    4   320 MARK       all      *      *       ::/0                 ::/0                 MARK or 0x2000
```

```
$> ip -6 route
::1 dev lo proto kernel metric 256 pref medium
2a00:1e:8085:6801:aaaa:fc80 dev eth0 proto kernel metric 100 pref medium
2a00:1e:8085:6801::/64 dev eth0 proto ra metric 100 pref medium
2a00:1e:8085:6801::/56 via fe80::aaaa:10b9 dev eth0 proto ra metric 100 pref high
fd8a:e26d:2709:9821::/64 dev podman1 proto kernel metric 256 pref medium
fe80::/64 dev podman1 proto kernel metric 256 pref medium
fe80::/64 dev veth0 proto kernel metric 256 pref medium
fe80::/64 dev eth0 proto kernel metric 1024 pref medium
default via fe80::aaaa:10b9 dev eth0 proto ra metric 100 pref high
```

```
$> ip a s
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000
    link/ether aa:aa:aa:aa:fc:80 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.252/24 brd 192.168.0.255 scope global dynamic noprefixroute enp1s0
       valid_lft 787406sec preferred_lft 787406sec
    inet6 2a00:1e:8085:6801:aaaa:fc80/128 scope global dynamic noprefixroute
       valid_lft 6124sec preferred_lft 2524sec
    inet6 fe80::aaaa:fc80/64 scope link noprefixroute
       valid_lft forever preferred_lft forever
43: podman1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether 9a:5f:ed:66:eb:a7 brd ff:ff:ff:ff:ff:ff
    inet 10.89.0.1/24 brd 10.89.0.255 scope global podman1
       valid_lft forever preferred_lft forever
    inet6 fd8a:e26d:2709:9821::1/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::88bd:2ff:fec4:93d3/64 scope link
       valid_lft forever preferred_lft forever
44: veth0@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue master podman1 state UP group default qlen 1000
    link/ether 9a:5f:ed:66:eb:a7 brd ff:ff:ff:ff:ff:ff link-netns netns-75d4a4b1-ec08-71e4-2df8-f470998156f7
    inet6 fe80::985f:edff:fe66:eba7/64 scope link
       valid_lft forever preferred_lft forever
```

# Introspection in container

```
$> ip a s
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0@if48: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether ae:cb:d3:a8:b4:26 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.89.0.24/24 brd 10.89.0.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fd8a:e26d:2709:9821::3/64 scope global
       valid_lft forever preferred_lft forever
    inet6 fe80::accb:d3ff:fea8:b426/64 scope link
       valid_lft forever preferred_lft forever
```

**Comment by Luap99** [comment]
I am by no means a ipv6 specialist but my understanding is that link-local address should never be routed anywhere. They are only valid on this specific link thus the name name link-local. So unless you have a direct link this cannot work.

**Comment by chrisbraucker** [comment]
Valid point, thank you. I'd say the question that comes out of this is whether the link-local addressed packet can be DNATed just like the GUA packets to leave on the right interface afterwards. The iptables rules netavark concocts seem to match the LL packets on the prerouting chain, but they seem to get lost somewhere later, and I fail to pinpoint this.
I assume that the LL packets need to be marked in prerouting and then handled with an extra route that acts on that mark.

**Comment by Luap99** [comment]
We have normal DNAT rules but I don't think they can work for link-local. The source address will not be changed which means the container interface cannot reach the source address, we would need to SNAT as well AFAIK and even then I don't know if it would work. In general I think it is expected for link-local to not be routed.

I would guess if you use the rootless networking tool like `--network pasta` or `--network slirp4netns` it could work because they are user space processes which actually bind the port on the host.

**Comment by sbrivio-rh** [comment]
By the way, for pasta, several users requested a functionality where we would track the prefix from a given host (via netlink monitor), and advertise updated, matching prefixes to the container.

At the moment, in that case, pasta implements an implicit NAT: the container gets a given source address (by default the first address of the upstream IPv6 interface), and will keep using that. When addresses change on the host, pasta uses native sockets without binding them to a specific address (again, by default), so the kernel will just start using whatever new address for us.

The gist of the feature request I mentioned is to avoid NAT altogether in that case.

**Comment by chrisbraucker** [comment]
@Luap99 Thank you for your responsiveness. The reason I worked with rootful containers is that I want to preserve as much of the origin information as possible (pihole/dns). With your input, and researching more about link local addresses I came to the same conclusion you summarized, so I'll close the issue.
Further testing led me to discover that rootful bridge networks in podman are fully routable from the outside as long as I advertise static routes on the local gateway and configure the firewall accordingly, which I will pursue for the time being.

@sbrivio-rh I've been testing with running dhclient sidecars in containers that use pasta, but since my ISP updates the public prefix very irregularly, testing has been slow. I currently lack the capacity to run this in a more mature environment.
I arrived at this setup because container forwards stopped being available when the initial address lease timed out, and so far it has been great. Pasta works very well for me, thank you!
